spring-security-oauth2-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/security/oauth2/resource/ResourceServerTokenServicesConfiguration.java

```diff
@@ -231,20 +241,50 @@ public DefaultTokenServices jwkTokenServices(TokenStore jwkTokenStore) {
 		@Bean
 		@ConditionalOnMissingBean(TokenStore.class)
 		public TokenStore jwkTokenStore() {
-			return new JwkTokenStore(this.resource.getJwk().getKeySetUri());
+			DefaultAccessTokenConverter accessTokenConverter = new DefaultAccessTokenConverter();
+			accessTokenConverter.setUserTokenConverter(userAuthenticationConverter());
+
+			return new JwkTokenStore(this.resource.getJwk().getKeySetUri(), accessTokenConverter);
+		}
+
+		@Override
+		public void setApplicationContext(ApplicationContext context) throws BeansException {
+			this.context = context;
+		}
+
+		private UserAuthenticationConverter userAuthenticationConverter() {
+			DefaultUserAuthenticationConverter userAuthenticationConverter =
+					new DefaultUserAuthenticationConverter();
+
+			if (hasExactlyOneBean(UserDetailsService.class)) {
+				UserDetailsService userDetailsService = this.context.getBean(UserDetailsService.class);
+				userAuthenticationConverter.setUserDetailsService(userDetailsService);
+			}
+
+			return userAuthenticationConverter;
+		}
+
+		private <T> boolean hasExactlyOneBean(Class<T> clazz) {
+			if (this.context == null) {
+				return false;
+			}
+
+			return this.context.getBeanNamesForType(clazz).length == 1;
 		}
 	}
 
 	@Configuration
 	@Conditional(JwtTokenCondition.class)
-	protected static class JwtTokenServicesConfiguration {
+	protected static class JwtTokenServicesConfiguration implements ApplicationContextAware {
```

Comment:
I would remove `implements ApplicationContextAware` in favour of `@Autowired`. This avoids the introduction of the public API `setApplicationContext `.